libafl/src/corpus/minimizer.rs

```diff
@@ -12,99 +12,97 @@ use num_traits::ToPrimitive;
 use z3::{ast::Bool, Config, Context, Optimize};
 
 use crate::{
-    bolts::AsIter,
+    bolts::{
+        tuples::{MatchName, Named},
+        AsIter,
+    },
     corpus::Corpus,
-    events::EventManager,
     executors::{Executor, HasObservers},
-    inputs::Input,
     observers::{MapObserver, ObserversTuple},
     schedulers::{LenTimeMulTestcaseScore, Scheduler, TestcaseScore},
-    state::{HasCorpus, HasMetadata},
-    Error, Evaluator, HasScheduler,
+    state::{HasCorpus, HasMetadata, UsesState},
+    Error, HasScheduler,
 };
 
 /// `CorpusMinimizers` minimize corpora according to internal logic. See various implementations for
 /// details.
-pub trait CorpusMinimizer<I, S>
+pub trait CorpusMinimizer<E>
 where
-    I: Input,
-    S: HasCorpus<I>,
+    E: UsesState,
+    E::State: HasCorpus,
 {
     /// Minimize the corpus of the provided state.
-    fn minimize<CS, EX, EM, OT, Z>(
+    fn minimize<CS, EM, Z>(
         &self,
         fuzzer: &mut Z,
-        executor: &mut EX,
+        executor: &mut E,
         manager: &mut EM,
-        state: &mut S,
+        state: &mut E::State,
     ) -> Result<(), Error>
     where
-        CS: Scheduler<I, S>,
-        EX: Executor<EM, I, S, Z> + HasObservers<I, OT, S>,
-        EM: EventManager<EX, I, S, Z>,
-        OT: ObserversTuple<I, S>,
-        Z: Evaluator<EX, EM, I, S> + HasScheduler<CS, I, S>;
+        E: Executor<EM, Z> + HasObservers,
+        CS: Scheduler<State = E::State>,
+        EM: UsesState<State = E::State>,
+        Z: HasScheduler<CS, State = E::State>;
 }
 
 /// Minimizes a corpus according to coverage maps, weighting by the specified `TestcaseScore`.
 ///
 /// Algorithm based on WMOPT: <https://hexhive.epfl.ch/publications/files/21ISSTA2.pdf>
 #[derive(Debug)]
-pub struct MapCorpusMinimizer<E, I, O, S, TS>
+pub struct MapCorpusMinimizer<E, O, T, TS>
 where
-    E: Copy + Hash + Eq,
-    I: Input,
-    for<'a> O: MapObserver<Entry = E> + AsIter<'a, Item = E>,
-    S: HasMetadata + HasCorpus<I>,
-    TS: TestcaseScore<I, S>,
+    E: UsesState,
+    E::State: HasCorpus + HasMetadata,
+    TS: TestcaseScore<E::State>,
 {
     obs_name: String,
-    phantom: PhantomData<(E, I, O, S, TS)>,
+    phantom: PhantomData<(E, O, T, TS)>,
 }
 
 /// Standard corpus minimizer, which weights inputs by length and time.
-pub type StdCorpusMinimizer<E, I, O, S> =
-    MapCorpusMinimizer<E, I, O, S, LenTimeMulTestcaseScore<I, S>>;
+pub type StdCorpusMinimizer<E, O, T> =
+    MapCorpusMinimizer<E, O, T, LenTimeMulTestcaseScore<<E as UsesState>::State>>;
 
-impl<E, I, O, S, TS> MapCorpusMinimizer<E, I, O, S, TS>
+impl<E, O, T, TS> MapCorpusMinimizer<E, O, T, TS>
 where
-    E: Copy + Hash + Eq,
-    I: Input,
-    for<'a> O: MapObserver<Entry = E> + AsIter<'a, Item = E>,
-    S: HasMetadata + HasCorpus<I>,
-    TS: TestcaseScore<I, S>,
+    E: UsesState,
+    E::State: HasCorpus + HasMetadata,
+    TS: TestcaseScore<E::State>,
```

Comment:
This type was completely incorrect. Apparently, when I wrote it, I got confused between E (for Entry, in MapObserver) and E (for Executor). As a result, it was deeply underconstrained and completely incorrect. The AT changes prevented the type from compiling and provided a very useful error message.